File: packages/battle_engine/test/turn_engine_test.dart
import 'package:battle_engine/battle_engine.dart';
import 'package:test/test.dart';

void main() {
  const playerA = Combatant(id: 'a', name: 'Ana');
  const playerB = Combatant(id: 'b', name: 'Beto');
  final engine = TurnEngine(defaultCombinationBook);

  group('TurnAction', () {
    test('throws if no elements are played', () {
      expect(
        () => TurnAction(actor: playerA, elements: const []),
        throwsArgumentError,
      );
    });
  });

  group('TurnEngine.playTurn', () {
    test('rejects an action from a combatant whose turn it is not', () {
      final state = BattleState.start(playerA: playerA, playerB: playerB);
      final action = TurnAction(actor: playerB, elements: [Elements.fire]);

      expect(() => engine.playTurn(state, action), throwsStateError);
    });

    test('passes the turn to the opponent after a valid action', () {
      final state = BattleState.start(playerA: playerA, playerB: playerB);
      final action = TurnAction(actor: playerA, elements: [Elements.fire]);

      final result = engine.playTurn(state, action);

      expect(result.state.currentTurn, equals(playerB));
    });

    test('a single played element never triggers a combination', () {
      final state = BattleState.start(playerA: playerA, playerB: playerB);
      final action = TurnAction(actor: playerA, elements: [Elements.fire]);

      final result = engine.playTurn(state, action);

      expect(result.triggeredCombination, isNull);
      expect(result.state.activeFieldEffects, isEmpty);
    });

    test('playing a known 2-element combination adds it to the field', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.fire, Elements.wind],
      );

      final result = engine.playTurn(state, action);

      expect(result.triggeredCombination?.resultId, equals('ignited_storm'));
      expect(result.state.activeFieldEffects, hasLength(1));
      expect(result.state.activeFieldEffects.first.id, equals('ignited_storm'));
    });

    test('playing an unknown combination advances the turn without adding '
        'a field effect', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.ice, Elements.shadow],
      );

      final result = engine.playTurn(state, action);

      expect(result.triggeredCombination, isNull);
      expect(result.state.activeFieldEffects, isEmpty);
      expect(result.state.currentTurn, equals(playerB));
    });

    test('turns alternate across multiple plays', () {
      var state = BattleState.start(playerA: playerA, playerB: playerB);

      state = engine
          .playTurn(
            state,
            TurnAction(actor: playerA, elements: [Elements.fire]),
          )
          .state;
      expect(state.currentTurn, equals(playerB));

      state = engine
          .playTurn(
            state,
            TurnAction(actor: playerB, elements: [Elements.water]),
          )
          .state;
      expect(state.currentTurn, equals(playerA));
    });

    test('field effects accumulate across turns', () {
      var state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {
          playerA: const ApPool(max: 5, current: 3),
          playerB: const ApPool(max: 5, current: 3),
        },
      );

      state = engine
          .playTurn(
            state,
            TurnAction(
              actor: playerA,
              elements: [Elements.fire, Elements.wind],
            ),
          )
          .state;

      state = engine
          .playTurn(
            state,
            TurnAction(
              actor: playerB,
              elements: [Elements.water, Elements.lightning],
            ),
          )
          .state;

      expect(state.activeFieldEffects, hasLength(2));
    });

    test('applies combinationModifiers to a triggered combination\'s '
        'field effect before adding it to the field', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.fire, Elements.wind],
      );

      final result = engine.playTurn(
        state,
        action,
        combinationModifiers: [CombinationModifiers.propagation],
      );

      expect(result.state.activeFieldEffects.single.area, equals(3));
    });

    test('combinationModifiers apply in order', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 5)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.earth, Elements.fire, Elements.water],
      );

      final result = engine.playTurn(
        state,
        action,
        combinationModifiers: [
          CombinationModifiers.propagation,
          CombinationModifiers.propagation,
        ],
      );

      expect(result.state.activeFieldEffects.single.area, equals(5));
    });

    test('combinationModifiers are ignored when no combination triggers', () {
      final state = BattleState.start(playerA: playerA, playerB: playerB);
      final action = TurnAction(actor: playerA, elements: [Elements.fire]);

      final result = engine.playTurn(
        state,
        action,
        combinationModifiers: [CombinationModifiers.propagation],
      );

      expect(result.state.activeFieldEffects, isEmpty);
    });

    test('with no combinationModifiers the combination result is '
        'unmodified', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.fire, Elements.wind],
      );

      final result = engine.playTurn(state, action);

      expect(result.state.activeFieldEffects.single.area, equals(1));
    });

    test('Tempestade deals 14 direct damage to the opponent', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.fire, Elements.wind],
      );

      final result = engine.playTurn(state, action);

      expect(result.state.hpOf(playerB).current, equals(86));
      expect(result.state.hpOf(playerA).current, equals(100));
    });

    test('the 3-element combination deals 35 damage', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 5)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.earth, Elements.fire, Elements.water],
      );

      final result = engine.playTurn(state, action);

      expect(result.state.hpOf(playerB).current, equals(65));
    });

    test('a single element deals 5 basic damage; an unknown combination '
        'deals none', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerB: const ApPool(max: 5, current: 3)},
      );

      final afterSingle = engine
          .playTurn(
            state,
            TurnAction(actor: playerA, elements: [Elements.fire]),
          )
          .state;
      expect(afterSingle.hpOf(playerB).current, equals(95));

      final afterUnknown = engine
          .playTurn(
            afterSingle,
            TurnAction(
              actor: playerB,
              elements: [Elements.ice, Elements.shadow],
            ),
          )
          .state;
      expect(afterUnknown.hpOf(playerA).current, equals(100));
    });

    test('Shield blocks the next combo damage and is consumed', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      ).withStatusApplied(playerB, ActiveStatus(effect: StatusEffects.shield));
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.fire, Elements.wind],
      );

      final result = engine.playTurn(state, action);

      expect(result.state.hpOf(playerB).current, equals(100));
      expect(result.state.hasStatus(playerB, StatusEffects.shield), isFalse);
    });

    test('Shield does not block a second hit after being consumed', () {
      var state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 4)},
      ).withStatusApplied(playerB, ActiveStatus(effect: StatusEffects.shield));

      state = engine
          .playTurn(
            state,
            TurnAction(
              actor: playerA,
              elements: [Elements.fire, Elements.wind],
            ),
          )
          .state; // blocked, shield consumed (4 seeded + 1 regen - 3 spent = 2 left)
      state = engine
          .playTurn(state, TurnAction(actor: playerB, elements: [Elements.ice]))
          .state; // no-op action, just passes the turn back
      state = engine
          .playTurn(
            state,
            TurnAction(
              actor: playerA,
              elements: [Elements.fire, Elements.wind],
            ),
          )
          .state; // 2 + 1 regen = 3, affordable again — not blocked this time

      expect(state.hpOf(playerB).current, equals(86));
    });

    test('a status with damagePerTick damages its owner at the end of '
        'every playTurn call, including the tick that expires it — on top '
        'of the actor\'s own basic damage', () {
      var state = BattleState.start(playerA: playerA, playerB: playerB)
          .withStatusApplied(
            playerB,
            ActiveStatus(
              effect: StatusEffects.burn,
              turnsRemaining: 2,
              damagePerTick: 8,
            ),
          );

      state = engine
          .playTurn(
            state,
            TurnAction(actor: playerA, elements: [Elements.fire]),
          )
          .state;
      // playerB: 100 - 5 (a's basic damage) - 8 (first DOT tick) = 87
      expect(state.hpOf(playerB).current, equals(87));

      state = engine
          .playTurn(
            state,
            TurnAction(actor: playerB, elements: [Elements.water]),
          )
          .state;
      // playerA: 100 - 5 (b's basic damage) = 95
      // playerB: 87 - 8 (second DOT tick, expires) = 79
      expect(state.hpOf(playerA).current, equals(95));
      expect(state.hpOf(playerB).current, equals(79));
      expect(state.hasStatus(playerB, StatusEffects.burn), isFalse);
    });

    test('combo damage can set a winner', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        playerBMaxHp: 14,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final result = engine.playTurn(
        state,
        TurnAction(actor: playerA, elements: [Elements.fire, Elements.wind]),
      );

      expect(result.state.hpOf(playerB).isDefeated, isTrue);
      expect(result.state.winner, equals(playerA));
    });

    test('DOT damage alone can set a winner', () {
      // playerBMaxHp is 10, not 5: a's basic damage (5) alone must not be
      // enough to defeat them — only the DOT tick (8) on top of it should.
      var state =
          BattleState.start(
            playerA: playerA,
            playerB: playerB,
            playerBMaxHp: 10,
          ).withStatusApplied(
            playerB,
            ActiveStatus(
              effect: StatusEffects.burn,
              turnsRemaining: 1,
              damagePerTick: 8,
            ),
          );

      final result = engine.playTurn(
        state,
        TurnAction(actor: playerA, elements: [Elements.ice]),
      );

      expect(result.state.winner, equals(playerA));
    });

    test('when DOT ticks would defeat both combatants in the same '
        'resolution, the actor wins the tie', () {
      // Both start at 8 HP (not 5): a's basic damage (5) to b alone must
      // not decide the winner ahead of the DOT tick this test is about —
      // it isolates the tie strictly to DOT tick ordering.
      final state =
          BattleState.start(
                playerA: playerA,
                playerB: playerB,
                playerAMaxHp: 8,
                playerBMaxHp: 8,
              )
              .withStatusApplied(
                playerA,
                ActiveStatus(
                  effect: StatusEffects.burn,
                  turnsRemaining: 1,
                  damagePerTick: 8,
                ),
              )
              .withStatusApplied(
                playerB,
                ActiveStatus(
                  effect: StatusEffects.burn,
                  turnsRemaining: 1,
                  damagePerTick: 8,
                ),
              );

      final result = engine.playTurn(
        state,
        TurnAction(actor: playerA, elements: [Elements.ice]),
      );

      expect(result.state.hpOf(playerA).isDefeated, isTrue);
      expect(result.state.hpOf(playerB).isDefeated, isTrue);
      expect(result.state.winner, equals(playerA));
    });

    test('playTurn throws once the battle already has a winner', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        playerBMaxHp: 1,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final finished = engine
          .playTurn(
            state,
            TurnAction(
              actor: playerA,
              elements: [Elements.fire, Elements.wind],
            ),
          )
          .state;
      expect(finished.winner, equals(playerA));

      expect(
        () => engine.playTurn(
          finished,
          TurnAction(actor: playerB, elements: [Elements.water]),
        ),
        throwsStateError,
      );
    });

    test('ends the match after enough basic (single-element) hits, with '
        'no AP involved', () {
      // 5 basic damage per hit — 20 hits defeat 100 HP. Actor A acts
      // first each round, so their 20th hit lands before B's 20th.
      var state = BattleState.start(playerA: playerA, playerB: playerB);
      for (var i = 0; i < 19; i++) {
        state = engine
            .playTurn(
              state,
              TurnAction(actor: playerA, elements: [Elements.fire]),
            )
            .state;
        state = engine
            .playTurn(
              state,
              TurnAction(actor: playerB, elements: [Elements.ice]),
            )
            .state;
      }
      final result = engine.playTurn(
        state,
        TurnAction(actor: playerA, elements: [Elements.fire]),
      );

      expect(result.state.hpOf(playerB).isDefeated, isTrue);
      expect(result.state.winner, equals(playerA));
    });
  });

  group('AP cost for combining elements', () {
    test('regenerates 1 AP for the actor at the start of their turn', () {
      final state = BattleState.start(playerA: playerA, playerB: playerB);
      final action = TurnAction(actor: playerA, elements: [Elements.fire]);

      final result = engine.playTurn(state, action);

      expect(result.state.apOf(playerA).current, equals(1));
      expect(result.state.apOf(playerB).current, equals(0));
    });

    test('AP regeneration is clamped at max', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 5)},
      );
      final action = TurnAction(actor: playerA, elements: [Elements.fire]);

      final result = engine.playTurn(state, action);

      expect(result.state.apOf(playerA).current, equals(5));
    });

    test('rejects a 2-element combination without enough AP', () {
      final state = BattleState.start(playerA: playerA, playerB: playerB);
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.fire, Elements.wind],
      );

      expect(() => engine.playTurn(state, action), throwsStateError);
    });

    test('rejects a 3-element combination that only affords a 2-element '
        'one', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.earth, Elements.fire, Elements.water],
      );

      expect(() => engine.playTurn(state, action), throwsStateError);
    });

    test('spends 3 AP on a successful 2-element combination', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 3)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.fire, Elements.wind],
      );

      final result = engine.playTurn(state, action);

      // seeded 3 + 1 regen = 4, minus 3 spent = 1
      expect(result.state.apOf(playerA).current, equals(1));
    });

    test('spends all 5 AP on a successful 3-element combination', () {
      final state = BattleState.start(
        playerA: playerA,
        playerB: playerB,
        ap: {playerA: const ApPool(max: 5, current: 5)},
      );
      final action = TurnAction(
        actor: playerA,
        elements: [Elements.earth, Elements.fire, Elements.water],
      );

      final result = engine.playTurn(state, action);

      expect(result.state.apOf(playerA).current, equals(0));
    });
  });
}

File: backend/test/battle-rules/turn-engine.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import {
  apOf,
  createBattleState,
  hasStatus,
  hpOf,
  withStatusApplied,
} from "../../src/battle-rules/battle-state.js";
import { defaultCombinationBook } from "../../src/battle-rules/combination-book.js";
import { propagation, volatility } from "../../src/battle-rules/combination-modifiers.js";
import { TurnValidationError } from "../../src/battle-rules/errors.js";
import { playTurn } from "../../src/battle-rules/turn-engine.js";

function startState() {
  return createBattleState({
    playerAId: "a",
    playerBId: "b",
    currentTurnId: "a",
    // Convenient default for tests that just need one combo to be
    // affordable right away — tests exercising the AP mechanic itself,
    // or needing more than one combo in a row, override this.
    ap: { a: { max: 5, current: 3 }, b: { max: 5, current: 3 } },
  });
}

test("throws if no elements are played", () => {
  assert.throws(
    () => playTurn(startState(), { actorId: "a", elementIds: [] }, defaultCombinationBook),
    TurnValidationError,
  );
});

test("rejects an action from a combatant whose turn it is not", () => {
  assert.throws(
    () =>
      playTurn(
        startState(),
        { actorId: "b", elementIds: ["fire"] },
        defaultCombinationBook,
      ),
    TurnValidationError,
  );
});

test("passes the turn to the opponent after a valid action", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire"] },
    defaultCombinationBook,
  );
  assert.equal(result.state.currentTurnId, "b");
});

test("a single played element never triggers a combination", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire"] },
    defaultCombinationBook,
  );
  assert.equal(result.triggeredCombinationId, null);
  assert.deepEqual(result.state.activeFieldEffects, []);
});

test("playing a known 2-element combination adds it to the field", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
  );
  assert.equal(result.triggeredCombinationId, "ignited_storm");
  assert.equal(result.state.activeFieldEffects.length, 1);
  assert.equal(result.state.activeFieldEffects[0]?.id, "ignited_storm");
});

test("playing an unknown combination advances the turn without adding a field effect", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["ice", "shadow"] },
    defaultCombinationBook,
  );
  assert.equal(result.triggeredCombinationId, null);
  assert.deepEqual(result.state.activeFieldEffects, []);
  assert.equal(result.state.currentTurnId, "b");
});

test("field effects accumulate across turns", () => {
  const first = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
  );
  const second = playTurn(
    first.state,
    { actorId: "b", elementIds: ["water", "lightning"] },
    defaultCombinationBook,
  );
  assert.equal(second.state.activeFieldEffects.length, 2);
});

test("a triggered combination damages the opponent, not the actor", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
  );
  assert.deepEqual(hpOf(result.state, "b"), { max: 100, current: 86 });
  assert.deepEqual(hpOf(result.state, "a"), { max: 100, current: 100 });
});

test("an unknown combination deals no damage", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["ice", "shadow"] },
    defaultCombinationBook,
  );
  assert.deepEqual(hpOf(result.state, "b"), { max: 100, current: 100 });
});

test("sets a winner once repeated basic (single-element) damage defeats a combatant", () => {
  // 5 basic damage per hit, always affordable (0 AP) — 20 hits defeat
  // 100 HP. Isolates "playTurn sets a winner"; combo/AP math is covered
  // by the AP-specific tests below.
  let state = startState();
  for (let i = 0; i < 19; i++) {
    state = playTurn(state, { actorId: "a", elementIds: ["fire"] }, defaultCombinationBook).state;
    state = playTurn(state, { actorId: "b", elementIds: ["ice"] }, defaultCombinationBook).state;
  }
  assert.equal(state.winner, null);
  const final = playTurn(
    state,
    { actorId: "a", elementIds: ["fire"] },
    defaultCombinationBook,
  );
  assert.equal(final.state.winner, "a");
});

test("throws when playing a turn after the battle is already over", () => {
  let state = startState();
  for (let i = 0; i < 19; i++) {
    state = playTurn(state, { actorId: "a", elementIds: ["fire"] }, defaultCombinationBook).state;
    state = playTurn(state, { actorId: "b", elementIds: ["ice"] }, defaultCombinationBook).state;
  }
  state = playTurn(state, { actorId: "a", elementIds: ["fire"] }, defaultCombinationBook).state;
  assert.equal(state.winner, "a");

  assert.throws(
    () => playTurn(state, { actorId: "b", elementIds: ["ice"] }, defaultCombinationBook),
    TurnValidationError,
  );
});

test("Shield blocks the next combo damage and is consumed", () => {
  const state = withStatusApplied(startState(), "b", {
    effectId: "shield",
    turnsRemaining: null,
    damagePerTick: 0,
  });

  const result = playTurn(
    state,
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
  );

  assert.equal(hpOf(result.state, "b").current, 100);
  assert.equal(hasStatus(result.state, "b", "shield"), false);
});

test("Shield does not block a second hit after being consumed", () => {
  let state = withStatusApplied(
    createBattleState({
      playerAId: "a",
      playerBId: "b",
      currentTurnId: "a",
      ap: { a: { max: 5, current: 4 } },
    }),
    "b",
    { effectId: "shield", turnsRemaining: null, damagePerTick: 0 },
  );

  state = playTurn(
    state,
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
  ).state; // blocked, shield consumed (4 seeded + 1 regen - 3 spent = 2 left)
  state = playTurn(
    state,
    { actorId: "b", elementIds: ["ice"] },
    defaultCombinationBook,
  ).state; // no-op, just passes the turn back
  state = playTurn(
    state,
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
  ).state; // 2 + 1 regen = 3, affordable again — not blocked this time

  assert.equal(hpOf(state, "b").current, 86);
});

test(
  "a status with damagePerTick damages its owner at the end of every " +
    "playTurn call, including the tick that expires it — on top of the " +
    "actor's own basic damage",
  () => {
    let state = withStatusApplied(startState(), "b", {
      effectId: "burn",
      turnsRemaining: 2,
      damagePerTick: 8,
    });

    state = playTurn(
      state,
      { actorId: "a", elementIds: ["fire"] },
      defaultCombinationBook,
    ).state;
    // b: 100 - 5 (a's basic damage) - 8 (first DOT tick) = 87
    assert.equal(hpOf(state, "b").current, 87);

    state = playTurn(
      state,
      { actorId: "b", elementIds: ["water"] },
      defaultCombinationBook,
    ).state;
    // a: 100 - 5 (b's basic damage) = 95
    // b: 87 - 8 (second DOT tick, expires) = 79
    assert.equal(hpOf(state, "a").current, 95);
    assert.equal(hpOf(state, "b").current, 79);
    assert.equal(hasStatus(state, "b", "burn"), false);
  },
);

test("DOT damage alone can set a winner", () => {
  // b's maxHp is 10, not 5: a's basic damage (5) alone must not be
  // enough to defeat them — only the DOT tick (8) on top of it should.
  const state = withStatusApplied(
    createBattleState({
      playerAId: "a",
      playerBId: "b",
      currentTurnId: "a",
      hp: { a: { max: 100, current: 100 }, b: { max: 10, current: 10 } },
    }),
    "b",
    { effectId: "burn", turnsRemaining: 1, damagePerTick: 8 },
  );

  const result = playTurn(
    state,
    { actorId: "a", elementIds: ["ice"] },
    defaultCombinationBook,
  );

  assert.equal(result.state.winner, "a");
});

test(
  "when DOT ticks would defeat both combatants in the same resolution, " +
    "the actor wins the tie",
  () => {
    // Both start at 8 HP (not 5): a's basic damage to b alone must not
    // decide the winner ahead of the DOT tick this test is about.
    let state = createBattleState({
      playerAId: "a",
      playerBId: "b",
      currentTurnId: "a",
      hp: { a: { max: 8, current: 8 }, b: { max: 8, current: 8 } },
    });
    state = withStatusApplied(state, "a", {
      effectId: "burn",
      turnsRemaining: 1,
      damagePerTick: 8,
    });
    state = withStatusApplied(state, "b", {
      effectId: "burn",
      turnsRemaining: 1,
      damagePerTick: 8,
    });

    const result = playTurn(
      state,
      { actorId: "a", elementIds: ["ice"] },
      defaultCombinationBook,
    );

    assert.equal(hpOf(result.state, "a").current, 0);
    assert.equal(hpOf(result.state, "b").current, 0);
    assert.equal(result.state.winner, "a");
  },
);

test(
  "applies combinationModifiers to a triggered combination's field effect " +
    "before adding it to the field",
  () => {
    const result = playTurn(
      startState(),
      { actorId: "a", elementIds: ["fire", "wind"] },
      defaultCombinationBook,
      [propagation],
    );

    assert.equal(result.state.activeFieldEffects[0]?.area, 3);
  },
);

test("combinationModifiers apply in order", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
    [propagation, propagation],
  );

  assert.equal(result.state.activeFieldEffects[0]?.area, 5);
});

test("combinationModifiers are ignored when no combination triggers", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire"] },
    defaultCombinationBook,
    [propagation],
  );

  assert.deepEqual(result.state.activeFieldEffects, []);
});

test("volatility reduces a triggered combination's duration", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
    [volatility],
  );

  // ignited_storm has a null (permanent) duration — volatility leaves it
  // unchanged, same as in battle_engine.
  assert.equal(result.state.activeFieldEffects[0]?.duration, null);
});

test("regenerates 1 AP for the actor at the start of their turn", () => {
  const state = createBattleState({ playerAId: "a", playerBId: "b", currentTurnId: "a" });
  const result = playTurn(state, { actorId: "a", elementIds: ["fire"] }, defaultCombinationBook);

  assert.deepEqual(apOf(result.state, "a"), { max: 5, current: 1 });
  assert.deepEqual(apOf(result.state, "b"), { max: 5, current: 0 });
});

test("AP regeneration is clamped at max", () => {
  const state = createBattleState({
    playerAId: "a",
    playerBId: "b",
    currentTurnId: "a",
    ap: { a: { max: 5, current: 5 } },
  });
  const result = playTurn(state, { actorId: "a", elementIds: ["fire"] }, defaultCombinationBook);

  assert.equal(apOf(result.state, "a").current, 5);
});

test("rejects a 2-element combination without enough AP", () => {
  const state = createBattleState({ playerAId: "a", playerBId: "b", currentTurnId: "a" });
  assert.throws(
    () => playTurn(state, { actorId: "a", elementIds: ["fire", "wind"] }, defaultCombinationBook),
    TurnValidationError,
  );
});

test("rejects a 3-element combination that only affords a 2-element one", () => {
  const state = createBattleState({
    playerAId: "a",
    playerBId: "b",
    currentTurnId: "a",
    ap: { a: { max: 5, current: 3 } },
  });
  assert.throws(
    () =>
      playTurn(
        state,
        { actorId: "a", elementIds: ["earth", "fire", "water"] },
        defaultCombinationBook,
      ),
    TurnValidationError,
  );
});

test("spends 3 AP on a successful 2-element combination", () => {
  const result = playTurn(
    startState(),
    { actorId: "a", elementIds: ["fire", "wind"] },
    defaultCombinationBook,
  );
  // startState seeds 3 + 1 regen = 4, minus 3 spent = 1
  assert.equal(apOf(result.state, "a").current, 1);
});

test("spends all 5 AP on a successful 3-element combination", () => {
  const state = createBattleState({
    playerAId: "a",
    playerBId: "b",
    currentTurnId: "a",
    ap: { a: { max: 5, current: 5 } },
  });
  const result = playTurn(
    state,
    { actorId: "a", elementIds: ["earth", "fire", "water"] },
    defaultCombinationBook,
  );
  assert.equal(apOf(result.state, "a").current, 0);
});
